# Base de Dados Escolhida para Estudo:
Ocorrências Aeronáuticas - ANAC
https://dados.gov.br/dados/conjuntos-dados/ocorrencias-aeronauticas

# Metadados:
https://www.anac.gov.br/acesso-a-informacao/dados-abertos/areas-de-atuacao/seguranca-operacional/ocorrencias-aeronauticas/76-ocorrencias-aeronauticas

# Dados Auxiliares
PDF: MCA 3-6 - Manual de Investigação do SIPAER
https://www2.fab.mil.br/cenipa/index.php/anexos/24-legsipaer

IDH:


# Descrição dos Atributos da Base de Dados
Campo:-									Descrição  
Número da Ocorrência:-					Número da Ocorrência Aeronáutica  
Classificação da Ocorrência:-			Classificação da Ocorrência Aeronáutica  
Data da Ocorrência:-					Data (UTC) da Ocorrência Aeronáutica  
Hora da Ocorrência:-					Hora (UTC) da Ocorrência Aeronáutica  
Município:-								Município do local da Ocorrência, quando aplicável  
UF:-									UF do local da Ocorrência, quando aplicável  
Latitude:-								Latitude do local onde houve a ocorrência  
Longitude:-								Longitude do local onde houve a ocorrência  
Tipo de Aeródromo:-						Tipo de Aeródromo em que houve a ocorrência, quando aplicável.  
​Histórico:-								Descrição sobre a ocorrência  
Matrícula:-								Matrícula da Aeronave associada à Ocorrência Aeronáutica  
Categoria da Aeronave:-					Categoria da Aeronave associada à Ocorrência Aeronáutica  
Operador:-								Operador da Aeronave associada à Ocorrência Aeronáutica  
Tipo de Ocorrência:-					Tipo de Ocorrência de acordo com a taxonomia ADREP  
Fase da Operação:-						Fase da Operação da Aeronave quando da Ocorrência Aeronáutica  
Operação:-								Tipo de Operação da Aeronave no momento da Ocorrência Aeronáutica  
Danos à Aeronave:-						Danos à Aeronave associada à Ocorrência Aeronáutica  
Aeródromo de Destino:-					Indicador de Localidade ICAO do Aeródromo de Destino da Aeronave associada à Ocorrência Aeronáutica  
Aeródromo de Origem:-					Indicador de Localidade ICAO do Aeródromo de Origem da Aeronave associada à Ocorrência Aeronáutica  
Lesões Fatais - Tripulantes:-			Número de Tripulantes com lesões fatais na Aeronave associada à Ocorrência Aeronáutica  
Lesões Fatais – Passageiros:-			Número de Passageiros com lesões fatais na Aeronave associada à Ocorrência Aeronáutica  
Lesões Fatais - Terceiros:-				Número de Terceiros com lesões fatais causadas pela Aeronave associada à Ocorrência Aeronáutica  
Lesões Graves – Tripulantes:-			Número de Tripulantes com lesões graves na Aeronave associada à Ocorrência Aeronáutica  
Lesões Graves – Passageiros:-			Número de Passageiros com lesões graves na Aeronave associada à Ocorrência Aeronáutica  
Lesões Graves – Terceiros:-				Número de Terceiros com lesões graves causadas pela Aeronave associada à Ocorrência Aeronáutica  
Lesões Leves - Tripulantes:-			Número de Tripulantes com lesões leves na Aeronave associada à Ocorrência Aeronáutica  
Lesões Leves – Passageiros:-			Número de Passageiros com lesões leves na Aeronave associada à Ocorrência Aeronáutica  
Lesões Leves - Terceiros:-				Número de Terceiros com lesões leves causadas pela Aeronave associada à Ocorrência Aeronáutica  
Ilesos – Tripulantes:-					Número de Tripulantes ilesos na Aeronave associada à Ocorrência Aeronáutica  
Ilesos – Passageiros:-					Número de Passageiros ilesos na Aeronave associada à Ocorrência Aeronáutica  
Lesões Desconhecidas – Tripulantes:-	Número de Tripulantes com lesões desconhecida na Aeronave associada à Ocorrência Aeronáutica  
Lesões Desconhecidas – Passageiros:-	Número de Passageiros com lesões desconhecida na Aeronave associada à Ocorrência Aeronáutica  
Lesões Desconhecidas - Terceiros:-		Número de Terceiros com lesões desconhecida causadas pela Aeronave associada à Ocorrência Aeronáutica  
PSSO:-									A ocorrência é considerada nos indicadores do Programa de Supervisão da Segurança Operacional (PSSO)  
Número da Ficha:-						Número da ficha cadastrada pelo CENIPA para a ocorrência.  
Operador Padronizado:-					Nome do operador/empresa, padronizado pelo sistema, responsável pela aeronave na época da ocorrência.  
Região:-								Região do País relacionada ao campo UF.  
ICAO:-									Código ICAO do aeródromo onde houve a ocorrência.  
Modelo:-								Descrição do Modelo da Aeronave  
CLS:-									Código que identifica o tipo da aeronave.  
Tipo ICAO:-								Código de identificação da aeronave segundo a ICAO.  
PMD:-									Peso máximo de Decolagem da aeronave  
Número de Assentos:-					Quantidade de assentos disponíveis na aeronave.  
Nome do Fabricante:-					Nome da empresa fabricante da aeronave.  
​Descrição Tipo:-						Descrição do tipo da ocorrência​  


# Recorte Temporal do Estudo
ano inicial:  
ano final:  
motivo:  

# Quais as Perguntas  
1- Quais os 3 tipos mais comuns de ocorrência por estado?  
2- Qual a média de ocorrências por região por ano?  
3- Quais fases das operações geram mais acidentes?  
4- Qual o total de fatalidades por operação por ano?  
5- Qual o tipo mais comum de ocorrência por categoria da aeronave?  
6- Qual a categoria de aeronave com menor indice de fatalidades? (total de fatalidades / total de ocorrências / número de acentos*)  
*normalização para comparação com demais aeronaves

# Importando Bibliotecas

In [1]:
import pandas as pd
import numpy as np
import requests
from io import StringIO

# Importando a base de dados do GitHub

In [ ]:
def carregar_csv_github(url_raw: str, sep: str = ",", skiprows: int = 0) -> pd.DataFrame:
    """
    Baixa um CSV público do GitHub e retorna como DataFrame pandas adaptável.
    """
    response = requests.get(url_raw, timeout=30)
    response.raise_for_status()

    # Lê o arquivo aplicando as configurações passadas nos argumentos
    df_lido = pd.read_csv(
        StringIO(response.text),
        sep=sep,
        skiprows=skiprows,
        on_bad_lines='skip'
    )

    # Transforma a string textual 'null' no valor NaN real do sistema
    df_limpo = df_lido.replace('null', np.nan)

    return df_limpo

# ==============================================================
# CARREGANDO E EXIBINDO A BASE DO CENIPA (Ocorrências)
# ==============================================================
csv_path_cenipa = "https://raw.githubusercontent.com/Jobanu/Projeto-Final-DS-PY-004/main/Dados/V_OCORRENCIA_AMPLA.csv"

# Carrega usando vírgula (sep=",") e sem pular linhas (skiprows=0)
df = carregar_csv_github(csv_path_cenipa, sep=";", skiprows=1)

print(f"--- BASE CENIPA: Carregadas {len(df)} ocorrências ---")
display(df.head())


# ==============================================================
# CARREGANDO E EXIBINDO A BASE DO IDH
# ==============================================================
csv_path_idhm = "https://raw.githubusercontent.com/Jobanu/Projeto-Final-DS-PY-004/desenvolvimento/Dados/IDHM.csv"

# Carrega usando ponto e vírgula (sep=";") e pulando a primeira linha (skiprows=1)
df_idhm = carregar_csv_github(csv_path_idhm, sep=";", skiprows=0)

print(f"\n--- BASE IDH: Carregadas {len(df_idhm)} linhas de estados ---")
display(df_idhm.head())


In [3]:
tamanho_df_inicio = len(df)
print(f"Tamanho inicial do DataFrame: {tamanho_df_inicio}")
memoria_df_inicio = df.memory_usage(deep=True).sum()
print(f"Tamanho inicial do DataFrame bytes: {memoria_df_inicio}")

Tamanho inicial do DataFrame: 5021
Tamanho inicial do DataFrame bytes: 11788464


# Explorando a base de dados

In [4]:
# Dimensões
df.shape

(5021, 45)

In [ ]:
# Visualizando as 5 primeiras linhas
df.head()

In [ ]:
# Visualizando as 5 últimas linhas
df.tail()

In [ ]:
# Visualizando os Metadados do Dataframe
df.info()

Atributos de [Data_da_Ocorrencia] e [Hora_da_Ocorrencia] carregados como object, é necessário ajustar este carregamento no comando de leitura do .csv:  
 4   Data_da_Ocorrencia                5021 non-null   object   
 5   Hora_da_Ocorrencia                3588 non-null   object  

 45 Colunas disponíveis. Será necessário selecionar as colunas próprias para o estudo, uma vez que fica inviável tratar todos dados nulls, vazios, outliners e depois não usar esses dados para o estudo.  
   

Será Necessário tratamento com cruzamento deste dados para preencimento do que for possível:  
Municipio                           1394  
UF                                     3  
Regiao                               670  

# Calculando NaN na Base de Dados Crua

In [8]:
def calcular_nan_por_atributo(df_NaN: pd.DataFrame) -> pd.DataFrame:
  # Calcula o número de valores NaN para cada coluna
  nan_counts = df_NaN.isnull().sum()

  # Calcula o total de registros no DataFrame
  total_records = len(df_NaN)

  # Cria um DataFrame com a contagem de NaN
  df_nan_analysis = pd.DataFrame({
      'Atributo': nan_counts.index,
      'NaN_Contagem': nan_counts.values
  })

  # Calcula o percentual de NaN's
  df_nan_analysis['NaN_Percentual'] = (df_nan_analysis['NaN_Contagem'] / total_records) * 100
  return df_nan_analysis

# Execução do código
df_nan_analysis = calcular_nan_por_atributo(df)
display(df_nan_analysis)

,Atributo,NaN_Contagem,NaN_Percentual
0,Numero_da_Ocorrencia,0,0.000000
1,Numero_da_Ficha,2,0.039833
2,Operador_Padronizado,76,1.513643
3,Classificacao_da_Ocorrencia,0,0.000000
4,Data_da_Ocorrencia,0,0.000000
5,Hora_da_Ocorrencia,1433,28.540131
6,Municipio,1394,27.763394
7,UF,3,0.059749
8,Regiao,670,13.343955
9,Descricao_do_Tipo,642,12.786298


In [9]:
df[df['UF'].isnull()][['UF', 'Historico']]

,UF,Historico
3766,NaN,Aircraft landed on RWY 24 and shortly after ma...
4705,NaN,A aeronave decolou do Aeródromo Formosa (SWFR)...
4928,NaN,"During cruise between WPT 0524N and KODOS, at ..."


É possível inferir que o atributo UF do ID 4705 é GO, os demais podem ser excluídos para o estudo de acidentes no território nacional, uma vez que, pelo histórico, ocorreram fora do território nacional.

In [10]:
df[df['Regiao'].isnull()][['UF', 'Historico']]

,UF,Historico
131,Exterior,"Immediately after take-off, with adverse meteo..."
132,Exterior,"Immediately after take-off, with adverse meteo..."
137,Exterior,A aeronave teve problemas no motor nº 1 durant...
369,Indeterminado,NaN
402,Indeterminado,"A aeronave decolou de local não registrado, co..."
...,...,...
4705,NaN,A aeronave decolou do Aeródromo Formosa (SWFR)...
4715,Exterior,Confidence Rating: Information is only availab...
4716,Exterior,Seaplane flooded during landing. Cockpit and f...
4851,Exterior,The Aircraft took off from Hong Kong Internati...


Podemos Excluir os registros com atributos [UF] == Exterior, uma vez que se trata de ocorrências fora do território nacional e, portanto, fora do range deste estudo que se restringe as ocorrências em território nacional.  
Podemos Excluir os registros com atributos [UF] == Indeterminado e [Histórico] == NaN, pois não dá para determinar o Estado da ocorrência para o estudo.  
Os demais Indeterminados, podem ser inferidos pelo histórico, mas devemos considerar o custo de uma Machine Learning para extrair esta informação ou fazer manualmente um a um.

Após o tratamento, devemos fazer a correlação [UF] x [Regiao] para preencher os dados faltantes de [Regiao] atráves dos dados existentes de [UF]

# Filtros do Estudo 
1 - [Numero_da_Ocorrencia] (tratando como indice)  
2 - [Classificacao_da_Ocorrencia]? 'Acidente' e/ou 'Incidente Grave'  
3 - Período do estudo baseado em [Data_da_Ocorrencia]  (Quais anos?)  
# Atributos de Interesse para as perguntas
1- agrupar por([UF]); contar([Tipo_de_Ocorrencia]); <-corresponde([Descricao_do_Tipo]); filtra(ano([Data_da_Ocorrencia]))  
  
2- agrupar por [Regiao]; agrupar por ano([Data_da_Ocorrencia]); média(soma([Lesoes_Fatais_Tripulantes]) soma([Lesoes_Fatais_Passageiros]) soma([Lesoes_Fatais_Terceiros]))  
  
3- agrupar por [Fase_da_Operacao]; contar([Numero_da_Ocorrencia])  
  
4- agrupar por [Fase_da_Operacao]; soma(soma([Lesoes_Fatais_Tripulantes]) soma([Lesoes_Fatais_Passageiros]) soma([Lesoes_Fatais_Terceiros])  
  
5- agrupar por [Categoria_da_Aeronave]; agrupar por [Tipo_de_Ocorrencia]. contar([Numero_da_Ocorrencia])  

6- agrupar por [Categoria_da_Aeronave]; 	soma(soma([Lesoes_Fatais_Tripulantes]) soma([Lesoes_Fatais_Passageiros]) soma([Lesoes_Fatais_Terceiros]) /
										contagem([Numero_da_Ocorrencia]) /
										soma([Numero_de_Assentos])


In [12]:
Atritubos_interesse = [
    'Numero_da_Ocorrencia',
    'Classificacao_da_Ocorrencia',
    'Data_da_Ocorrencia',
    'Historico',
    'UF',
    'Regiao',
    'Tipo_de_Ocorrencia',
    'Descricao_do_Tipo',
    'Operacao',
    'Lesoes_Fatais_Tripulantes',
    'Lesoes_Fatais_Passageiros',
    'Lesoes_Fatais_Terceiros',
    'Lesoes_Graves_Tripulantes',
    'Lesoes_Graves_Passageiros',
    'Lesoes_Graves_Terceiros',
    'Lesoes_Leves_Tripulantes',
    'Lesoes_Leves_Passageiros',
    'Lesoes_Leves_Terceiros',
    'Ilesos_Tripulantes',
    'Ilesos_Passageiros',
    'Lesoes_Desconhecidas_Tripulantes',
    'Lesoes_Desconhecidas_Passageiros',
    'Lesoes_Desconhecidas_Terceiros',
    'Fase_da_Operacao',
    'Categoria_da_Aeronave',
    'Numero_de_Assentos'
    ]

Será necessário estudar o forma de preencher os dados Nulls nestes Atributos, pois são atributos de interesse no estudo:  
Lesoes_Fatais_Tripulantes           1118  
Lesoes_Fatais_Passageiros           1239  
Lesoes_Fatais_Terceiros             1319  
Lesoes_Graves_Tripulantes           1168  
Lesoes_Graves_Passageiros           1281  
Lesoes_Graves_Terceiros             1327  
Lesoes_Leves_Tripulantes            1124  
Lesoes_Leves_Passageiros            1277  
Lesoes_Leves_Terceiros              1327  
Ilesos_Tripulantes                   410  
Ilesos_Passageiros                   949  
Lesoes_Desconhecidas_Tripulantes    1315  
Lesoes_Desconhecidas_Passageiros    1326  
Lesoes_Desconhecidas_Terceiros      1327  

In [ ]:
# 1. Lista com as colunas de lesões e contagem da base
colunas_lesoes = [
    'Lesoes_Fatais_Tripulantes', 'Lesoes_Fatais_Passageiros', 'Lesoes_Fatais_Terceiros',
    'Lesoes_Graves_Tripulantes', 'Lesoes_Graves_Passageiros', 'Lesoes_Graves_Terceiros',
    'Lesoes_Leves_Tripulantes', 'Lesoes_Leves_Passageiros', 'Lesoes_Leves_Terceiros',
    'Ilesos_Tripulantes', 'Ilesos_Passageiros',
    'Lesoes_Desconhecidas_Tripulantes', 'Lesoes_Desconhecidas_Passageiros', 'Lesoes_Desconhecidas_Terceiros'
]

df[colunas_lesoes].head(10)

# Relendo o CSV com os novos filtros

In [35]:
# Importando a base de dados do GitHub com filtros
def carregar_csv_github2(url_raw: str, pula_linha = 1) -> pd.DataFrame:
    """
    Baixa um CSV público do GitHub e retorna como DataFrame pandas.
    url_raw: Caminho no GitHub para o arquivo .CSV
    """
    response = requests.get(url_raw, timeout=30)
    response.raise_for_status()

    # 1. Lê o arquivo agora com somente os atributos de interesse
    # e com configuração de correta da [Data_da_Ocorrencia]
    df_lido = pd.read_csv(
        StringIO(response.text),
        sep=';',
        skiprows=pula_linha,
        on_bad_lines='skip',
        usecols=Atritubos_interesse,
        parse_dates=['Data_da_Ocorrencia']
    )

    # 2. Transforma a string textual 'null' no valor NaN real do sistema
    df_limpo = df_lido.replace('null', np.nan)

    # 3. Retorna o DataFrame perfeitamente estruturado
    return df_limpo

# Execução do código
#df = carregar_csv_github2(csv_path)
df = "https://raw.githubusercontent.com/Jobanu/Projeto-Final-DS-PY-004/main/Dados/V_OCORRENCIA_AMPLA.csv"

print(f"Carregadas {len(df)} ocorrências")

Carregadas 98 ocorrências


# Verificando as novas dimensões do Dataframe

In [15]:
df.shape

(5021, 45)

In [ ]:
df.head()

In [ ]:
df.info()

In [18]:
print(df.isna().sum())# Execução do código
df_nan_analysis_novo = calcular_nan_por_atributo(df)
display(df_nan_analysis_novo)

Numero_da_Ocorrencia                   0
Numero_da_Ficha                        2
Operador_Padronizado                  76
Classificacao_da_Ocorrencia            0
Data_da_Ocorrencia                     0
Hora_da_Ocorrencia                  1433
Municipio                           1394
UF                                     3
Regiao                               670
Descricao_do_Tipo                    642
ICAO                                3334
Latitude                            1490
Longitude                           1490
Tipo_de_Aerodromo                   3449
Historico                            265
Matricula                              4
Categoria_da_Aeronave                  9
Operador                              76
Tipo_de_Ocorrencia                   642
Fase_da_Operacao                    1300
Operacao                              94
Danos_a_Aeronave                    1426
Aerodromo_de_Destino                2138
Aerodromo_de_Origem                 2117
Lesoes_Fatais_Tr

,Atributo,NaN_Contagem,NaN_Percentual
0,Numero_da_Ocorrencia,0,0.000000
1,Numero_da_Ficha,2,0.039833
2,Operador_Padronizado,76,1.513643
3,Classificacao_da_Ocorrencia,0,0.000000
4,Data_da_Ocorrencia,0,0.000000
5,Hora_da_Ocorrencia,1433,28.540131
6,Municipio,1394,27.763394
7,UF,3,0.059749
8,Regiao,670,13.343955
9,Descricao_do_Tipo,642,12.786298


# Iniciando o tratamento dos Dados

## Lesões

## Tratamento das colunas dos tipos de Lesões - (NaN para 0)

A razão para escolhermos apenas as colunas de lesões e não o DataFrame inteiro é que o significado do "vazio" (NaN) muda completamente dependendo do tipo de dado da coluna. Substituir tudo por zero de forma automática geraria graves erros de lógica e distorções na análise.

1. Colunas de Texto (Strings) ficariam sem sentido

    - Operador: O responsável pelo voo passaria a se chamar 0 (o Pandas não saberia que ali deveria ser "Desconhecido")
    Matricula: O prefixo do avião viraria 0.

2. Colunas Numéricas Específicas seriam corrompidas, pois onúmero 0 possui um valor matemático real em engenharia aeronáutica.

    - PMD (Peso Máximo de Decolagem): Se uma aeronave experimental não teve o peso registrado, colocar 0 faria o Pandas calcular que o avião pesa zero quilos, o que estragaria qualquer média de peso que você tentasse fazer no projeto.
    - Numero_de_Assentos: Colocar 0 indicaria que o avião voava sem nenhum banco dentro, quando na verdade o dado apenas não foi coletado.

3. Nas colunas de Lesões, o 0 mantém a lógica perfeitaNas colunas de contagem de pessoas (Lesoes_Fatais, Lesoes_Graves), assumir que o vazio equivale a 0 é seguro e correto para o modelo de dados, pois:

    - Se o relatório não mencionou feridos, a hipótese estatística padrão é que zero pessoas se machucaram naquela categoria específica.
    - Permite converter a coluna de Float (número quebrado) para Int (número inteiro), já que não existem "1.5 pessoas mortas".

In [ ]:
# 1. Lista com as colunas de lesões e contagem da base
colunas_lesoes = [
    'Lesoes_Fatais_Tripulantes', 'Lesoes_Fatais_Passageiros', 'Lesoes_Fatais_Terceiros',
    'Lesoes_Graves_Tripulantes', 'Lesoes_Graves_Passageiros', 'Lesoes_Graves_Terceiros',
    'Lesoes_Leves_Tripulantes', 'Lesoes_Leves_Passageiros', 'Lesoes_Leves_Terceiros',
    'Ilesos_Tripulantes', 'Ilesos_Passageiros',
    'Lesoes_Desconhecidas_Tripulantes', 'Lesoes_Desconhecidas_Passageiros', 'Lesoes_Desconhecidas_Terceiros'
]

# --- ANTES ---
print("=== 1. QUANTIDADE DE NaNs ANTES DA LIMPEZA ===")
# Soma quantos NaNs existem apenas nas colunas da lista
print(df[colunas_lesoes].isna().sum())
print("-" * 50)


# 2. Criar a função de preenchimento
def preencher_nan_lesoes(dataframe: pd.DataFrame, lista_colunas: list) -> pd.DataFrame:
    """
    Substitui os valores NaN das colunas de lesões pelo número zero (0).
    Aplica a conversão para tipo inteiro (Int64), ideal para contagens.
    """
    # Faz uma cópia para não alterar o DataFrame original por acidente
    df_copia = dataframe.copy()
    
    # Preenche os NaNs com 0 e converte para tipo numérico inteiro
    df_copia[lista_colunas] = df_copia[lista_colunas].fillna(0).astype('int64')
    
    return df_copia


# 3. Aplicar a função no seu DataFrame
df = preencher_nan_lesoes(df, colunas_lesoes)


# --- DEPOIS ---
print("=== 2. QUANTIDADE DE NaNs DEPOIS DA LIMPEZA ===")
print(df[colunas_lesoes].isna().sum())
print("-" * 50)



## Valores únicos

In [ ]:
# Lista para armazenar os dados dos valores únicos de cada coluna
dados_valores_unicos = []

# Itera sobre cada coluna do DataFrame df
for coluna in df.columns:
    valor_unico = df[coluna].unique()
    numero_valores_unicos = len(valor_unico)
    dados_valores_unicos.append({
        'Coluna': coluna,
        'Valores_Unicos': valor_unico.tolist(), # Converte array numpy para lista
        'Numero_de_Valores_Unicos': numero_valores_unicos
    })

# Cria um DataFrame a partir dos dados coletados
df_valores_unicos = pd.DataFrame(dados_valores_unicos)

print('==== Resumo de Valores Únicos por Atributo no df ====\n')
# Exibe o DataFrame resultante
display(df_valores_unicos)

## [Numero_de_Assentos]

In [ ]:
df['Numero_de_Assentos'].isna().sum()

Trantando o [Numero_de_assentos] pela média existente calculada pela [Categoria_da_Aeronave]

In [ ]:
media_assentos_por_categoria = df.groupby('Categoria_da_Aeronave')['Numero_de_Assentos'].mean().reset_index()
media_assentos_por_categoria = media_assentos_por_categoria.dropna(subset=['Numero_de_Assentos'])
media_assentos_por_categoria['Numero_de_Assentos'] = media_assentos_por_categoria['Numero_de_Assentos'].astype('int64')
display(media_assentos_por_categoria)

In [ ]:
# Criar um dicionário de mapeamento de Categoria_da_Aeronave para Numero_de_Assentos médios
assentos_medios_dict = media_assentos_por_categoria.set_index('Categoria_da_Aeronave')['Numero_de_Assentos'].to_dict()

# Mapear os valores médios para a coluna 'Numero_de_Assentos' do df
# Usar apply com uma função lambda para preencher apenas NaNs
df['Numero_de_Assentos'] = df.apply(
    lambda row: assentos_medios_dict.get(row['Categoria_da_Aeronave'], row['Numero_de_Assentos'])
    if pd.isna(row['Numero_de_Assentos']) else row['Numero_de_Assentos'],
    axis=1
)

# Converter a coluna para int64 após o preenchimento, se necessário (lidando com NaNs remanescentes, se houver)
df['Numero_de_Assentos'] = df['Numero_de_Assentos'].fillna(0).astype('int64')

print(f"NaNs restantes em 'Numero_de_Assentos': {df['Numero_de_Assentos'].isnull().sum()}")
display(df[['Categoria_da_Aeronave', 'Numero_de_Assentos']].head())

## [UF] e [Região]

### [UF]

In [ ]:
df['UF'].isna().sum()

In [ ]:
df[df['UF'].isnull()][['Numero_da_Ocorrencia', 'UF', 'Historico']]

Com o [Numero_da_Ocorrencia] 46147 já tratado, vamos agora remover as ocorrências onde a [UF] ainda está como nula. Conforme a análise anterior, estas são ocorrências fora do território nacional e não são relevantes para o estudo.

In [ ]:
#Conforme Histórico, ID 4705 tem a UF em GO
df.loc[df['Numero_da_Ocorrencia'] == 46147, 'UF'] = 'GO'
display(df[df['UF'].isnull()][['Numero_da_Ocorrencia', 'UF', 'Historico']])

In [ ]:
df.dropna(subset=['UF'], inplace=True)
print(f"NaNs restantes em 'UF': {df['UF'].isnull().sum()}")

### [Regiao]

Vamos preencher os valores ausentes no atributo [Regiao] do df. Para isso, usaremos o dicionário regioes_brasil para mapear cada [UF] à sua Regiao correspondente.

In [ ]:
# Dicionário de Regiões do Brasil
regioes_brasil = {
    "Norte": ["AC", "AM", "AP", "PA", "RO", "RR", "TO"],
    "Nordeste": ["AL", "BA", "CE", "MA", "PB", "PE", "PI", "RN", "SE"],
    "Centro-Oeste": ["DF", "GO", "MT", "MS"],
    "Sudeste": ["ES", "MG", "RJ", "SP"],
    "Sul": ["PR", "RS", "SC"]
}


In [ ]:
# Inverter o dicionário para mapear UF para Região
uf_para_regiao = {
    uf: regiao
    for regiao, ufs in regioes_brasil.items()
    for uf in ufs
}

# Criar um Series com as novas regiões mapeadas a partir da UF
# Valores de UF que não estão no dicionário resultarão em NaN aqui
regioes_mapeadas = df['UF'].map(uf_para_regiao)

# Preencher APENAS os NaN em 'Regiao' com os valores de 'regioes_mapeadas'
# O aviso FuturegWarning é resolvido ao atribuir o resultado de volta à coluna
df['Regiao'] = df['Regiao'].fillna(regioes_mapeadas)

print(f"NaNs restantes em 'Regiao': {df['Regiao'].isnull().sum()}")

Vamos investigar quais valores de [UF] não foram mapeados e ainda resultam em NaN na coluna [Regiao].

In [ ]:
ufs_nao_mapeadas = df[df['Regiao'].isnull()]['UF'].unique()
print(f"UFs que não foram mapeadas para 'Regiao': {ufs_nao_mapeadas.tolist()}")

Vamos proceder à remoção dos registros contendo 'Exterior', pois não faram parte do estudo

In [ ]:
# Remove as linhas onde 'UF' == 'Exterior'
df = df[~df['UF'].isin(['Exterior'])]

# Verifica novamente os NaNs na coluna 'Regiao' após a remoção
print(f"NaNs restantes em 'Regiao' após a remoção de 'Exterior': {df['Regiao'].isnull().sum()}")

Vamos remover os registros onde a [UF] é 'Indeterminado' e o [Historico] está NaN.

In [ ]:
# Remove as linhas onde 'UF' é 'Indeterminado' E 'Historico' é NaN
df = df[~((df['UF'] == 'Indeterminado') & (df['Historico'].isnull()))]

# Verifica novamente os NaNs na coluna 'Regiao' após a remoção e re-mapeamento
print(f"NaNs restantes em 'Regiao' após a remoção de 'Indeterminado' com Historico NaN: {df['Regiao'].isnull().sum()}")

Como não está no escopo do estudo a utilização de Machine Learning, IA Generativa ou Agentes de IA para tratamento dos dados, usando o [Historico] como dado para inferencia da [UF], vamos excluir os demais Indeterminados para continuação do estudo.  
A mesma decisão será usada nas demais análises.

In [ ]:
# Remove as linhas onde 'UF' == 'Indeterminado'
df = df[~df['UF'].isin(['Indeterminado'])]

# Verifica novamente os NaNs na coluna 'Regiao' após a remoção
print(f"NaNs restantes em 'Regiao' após a remoção de 'Indeterminado': {df['Regiao'].isnull().sum()}")

## [Fase_da_Operacao]

In [ ]:
display(df[df['Fase_da_Operacao'].isnull()][['Fase_da_Operacao', 'Descricao_do_Tipo', 'Historico']])

Não é possível inferir uma [Fase_da_Operação] a partir dos dados que temos para preenchimento dos registros com NaN, portanto, optamos por excluir do estudo estes registros.

In [ ]:
df.dropna(subset=['Fase_da_Operacao'], inplace=True)
print(f"NaNs restantes em 'Fase_da_Operacao': {df['Fase_da_Operacao'].isnull().sum()}")

## [Tipo_de_Ocorrencia]

In [ ]:
df_tp_null = df[df['Tipo_de_Ocorrencia'].isnull()][['Numero_da_Ocorrencia', 'Tipo_de_Ocorrencia', 'Descricao_do_Tipo', 'Historico']]
df_tp_null.to_csv('df_tp_null.csv', index=False)
df_tp_null

Como temos 20 ocorrências, vamos analisar e atualizar manualmente os dados de tipo de ocorrência, de acordo com a tabela de Tipos de Ocorrências do 'MCA 3-6 - Manual de Investigação do SIPAER' - Anexo C, pags. 359 - 380.  
Faremos os acertos usando o df_tp_null.csv criado, retornando com um o tp_null_acertado.csv com os valores a serem corrigidos.


In [ ]:
# Importando a base de dados do GitHub
def carregar_csv_github(url_raw: str, pula_linha = 1) -> pd.DataFrame:
    """
    Baixa um CSV público do GitHub e retorna como DataFrame pandas.
    url_raw: Caminho no GitHub para o arquivo .CSV
    """
    response = requests.get(url_raw, timeout=30)
    response.raise_for_status()

    # 1. Lê o arquivo pulando os metadados do topo
    df_lido = pd.read_csv(
        StringIO(response.text),
        sep=',',
        on_bad_lines='skip'
    )

    # 2. Transforma a string textual 'null' no valor NaN real do sistema
    df_limpo = df_lido.replace('null', np.nan)

    # 3. Retorna o DataFrame perfeitamente estruturado
    return df_limpo

# Execução do código
tp_null_path = "https://raw.githubusercontent.com/Jobanu/Projeto-Final-DS-PY-004/main/Dados/tp_null_acertado.csv"
df_tp = carregar_csv_github(tp_null_path)

print(f"Carregadas {len(df_tp)} ocorrências")


Com base nos dados corrigidos do df_tp contendo os acertos manuais, vamos atualizar as colunas [Tipo_de_Ocorrencia] e [Descricao_do_Tipo] no df para preencher os NaN's.

In [ ]:
# Cria dicionários de mapeamento a partir de df_tp
tipo_ocorrencia_map = df_tp.set_index('Numero_da_Ocorrencia')['Tipo_de_Ocorrencia'].to_dict()
descricao_tipo_map = df_tp.set_index('Numero_da_Ocorrencia')['Descricao_do_Tipo'].to_dict()

# Identifica as linhas em df onde 'Tipo_de_Ocorrencia' é NaN
mask_tipo_nan = df['Tipo_de_Ocorrencia'].isnull()
# Identifica as linhas em df onde 'Descricao_do_Tipo' é NaN
mask_descricao_nan = df['Descricao_do_Tipo'].isnull()

# Preenche os NaNs em 'Tipo_de_Ocorrencia' usando o mapeamento
df.loc[mask_tipo_nan, 'Tipo_de_Ocorrencia'] = df.loc[mask_tipo_nan, 'Numero_da_Ocorrencia'].map(tipo_ocorrencia_map)

# Preenche os NaNs em 'Descricao_do_Tipo' usando o mapeamento
df.loc[mask_descricao_nan, 'Descricao_do_Tipo'] = df.loc[mask_descricao_nan, 'Numero_da_Ocorrencia'].map(descricao_tipo_map)

# Verifica quantos NaNs ainda existem na coluna 'Tipo_de_Ocorrencia'
remaining_nan_tipo = df['Tipo_de_Ocorrencia'].isnull().sum()
print(f"NaNs restantes em 'Tipo_de_Ocorrencia' após a atualização: {remaining_nan_tipo}")

As demais ocorrências seram apagadas, pois não é possível determinar estes dados.

In [ ]:
df.dropna(subset=['Tipo_de_Ocorrencia'], inplace=True)
print(f"NaNs restantes em 'Tipo_de_Ocorrencia': {df['Tipo_de_Ocorrencia'].isnull().sum()}")

### [Categoria_da_Aeronave]

In [ ]:
df[df['Categoria_da_Aeronave'].isnull()][['Numero_da_Ocorrencia', 'Categoria_da_Aeronave', 'Historico']]

Não é possível definir a [Categoria_da_Aeronave] pelo [Historico], portanto, vamos excluir as linhas não preenchidas.


In [ ]:
df.dropna(subset=['Categoria_da_Aeronave'], inplace=True)
print(f"NaNs restantes em 'Categoria_da_Aeronave': {df['Categoria_da_Aeronave'].isnull().sum()}")

## [operacao]

In [ ]:
df[df['Operacao'].isnull()][['Numero_da_Ocorrencia', 'Operacao', 'Historico']]

Conforme analise dos dados únicos existe a possíbilidade de preenchimento como 'Desconhecida', portanto, vamos trocar os 13 registros NaN's por 'Desconhecida' para não perdermos os demais dados utéis destes regsitros.

In [ ]:
df['Operacao'].fillna('Desconhecida', inplace=True)
print(f"NaNs restantes em 'Operacao': {df['Operacao'].isnull().sum()}")

### Consultando as informações do DataFrame para verificação do sucesso da limpeza de NaN's

In [ ]:
df.info()

### [Numero_da_Ocorrencia]

Ocorrências duplicadas

In [ ]:

# Calcula as ocorrencias em 'Numero_da_Ocorrencia'
conta_ocorrencias = df['Numero_da_Ocorrencia'].value_counts()

# Filtra valores que aparecem mais de uma vez
ocorrencias_repetidas = conta_ocorrencias[conta_ocorrencias > 1]

# Converte o resultado para NumPy array
ocorrencias_repetidas_array = np.array([
    [num_oc, contagem] for num_oc, contagem in ocorrencias_repetidas.items()
])

print("Valores repetidos no atributo 'Numero_da_Ocorrencia' e suas contagens:")
print(ocorrencias_repetidas_array)

Verificando a relação de tipos de Ocorrências nos Números de ocorrências repetidos.  
Queremos saber se as ocorrências repetidas estão relacionanda a tipos de ocorrências diferentes, ou seja, possíveis agentes diferentes para mesma ocorrência ou erro de registro realmente duplicado.

In [ ]:
# Extrai os números de ocorrência que se repetem, a partir da análise
# anterior
numeros_ocorrencias_duplicadas = ocorrencias_repetidas.index.tolist()

# Lista para armazenar os resultados
lista_resultados_tipo_ocorrencia = []

# Itera sobre cada número de ocorrência que foi detectado como duplicado
for num_ocorrencia in numeros_ocorrencias_duplicadas:
    # Filtra o DataFrame original (df) para este número de ocorrência
    df_filtrado_por_num = df[df['Numero_da_Ocorrencia'] == num_ocorrencia]

    # Coleta todos os tipos de ocorrência únicos para este número de ocorrência
    tipos_de_ocorrencia_para_este_num = df_filtrado_por_num['Tipo_de_Ocorrencia'].unique()

    # Verifica se há mais de um tipo de ocorrência único para este número de ocorrência
    # Se houver mais de um, significa que o Tipo_de_Ocorrencia se repete (ou seja, varia)
    qtd_tipos_unicos = len(df_filtrado_por_num)
    tipo_de_ocorrencia_varia_nesta_repeticao = len(tipos_de_ocorrencia_para_este_num) > 1

    # Adiciona o resultado à lista: [número da ocorrência, lista de tipos únicos, se o tipo varia]
    lista_resultados_tipo_ocorrencia.append(
        [num_ocorrencia, tipos_de_ocorrencia_para_este_num.tolist(), tipo_de_ocorrencia_varia_nesta_repeticao, qtd_tipos_unicos]
    )

# Converte a lista de resultados em um array NumPy
# Usamos dtype=object para permitir que as listas de tipos de ocorrência tenham tamanhos variáveis
array_verificacao_tipos = np.array(lista_resultados_tipo_ocorrencia, dtype=object)

print("=== Verificação de 'Tipo_de_Ocorrencia' para 'Numero_da_Ocorrencia' repetidos ===")
print("Colunas: [Número da Ocorrência, Tipos Únicos de Ocorrência, TipSo de Ocorrência varia?]")
print(array_verificacao_tipos)

print('-' * 50)
print('Ocorrencias Verdadeiramente Dupllicadas')



Executando as exclusões com combinação entre [Numero_da_Ocorrencia] & [Tipo_de_Ocorrencia]

In [ ]:
#Analise Antes da Execução do código
print(f"Quantidade de linhas antes da remoção de duplicatas: {len(df)}")
display(df[df['Numero_da_Ocorrencia'] == 36228])

# Remove duplicatas com base na combinação de 'Numero_da_Ocorrencia' e 'Tipo_de_Ocorrencia'
df.drop_duplicates(subset=['Numero_da_Ocorrencia', 'Tipo_de_Ocorrencia'], inplace=True)

#Analise Após a Execução do código
print(f"Quantidade de linhas após a remoção de duplicatas: {len(df)}")
display(df[df['Numero_da_Ocorrencia'] == 36228])

## Verificando as novas dimensões do DataFrame

In [ ]:
tamanho_df_final = len(df)
print("Quantidade de Registros:")
print(f'Inicial do DataFrame: {tamanho_df_inicio} Registros')
print(f"Final do DataFrame: {tamanho_df_final} Registros")
print(f'Diferença: {tamanho_df_final - tamanho_df_inicio} Registros')

print("-" * 50)

memoria_df_final = df.memory_usage(deep=True).sum()
print("Quantidade de Memória:")
print(f"Inicial do DataFrame: {memoria_df_inicio/1000000} Mb")
print(f"Final do DataFrame: {memoria_df_final/1000000} Mb")
print(f'Diferença: {(memoria_df_final - memoria_df_inicio)/1000000} Mb')

print("-" * 50)

print('Percentual de Aproveitamento de Dados')
print(f'Percentual de Registros Aproveitados: {(tamanho_df_final / tamanho_df_inicio) * 100}%')
print(f'Percentual de Redução no Uso da Memória: {(1 - (memoria_df_final / memoria_df_inicio)) * 100}%')

## Analise Estatística Descritiva Geral

In [ ]:
df.describe()

## Criando Atributo com o Ano da Ocorrência

In [ ]:
df['Data_da_Ocorrencia'] = pd.to_datetime(df['Data_da_Ocorrencia'],format='%Y-%m-%d')
df['Ano_da_Ocorrencia'] = df['Data_da_Ocorrencia'].dt.year
df.info()
display(df)

## Criando a Totalização de Lesões por Fatais, Graves, Leves, Desconhecidas, e Ilesos

In [ ]:
df[colunas_lesoes].columns

In [ ]:
df['Total_Lesoes_Fatais'] = df['Lesoes_Fatais_Tripulantes'] + df['Lesoes_Fatais_Passageiros'] + df['Lesoes_Fatais_Terceiros']
df['Total_Lesoes_Graves'] = df['Lesoes_Graves_Tripulantes'] + df['Lesoes_Graves_Passageiros'] + df['Lesoes_Graves_Terceiros']
df['Total_Lesoes_Leves'] = df['Lesoes_Leves_Tripulantes'] + df['Lesoes_Leves_Passageiros'] + df['Lesoes_Leves_Terceiros']
df['Total_Lesoes_Desconhecidas'] = df['Lesoes_Desconhecidas_Tripulantes'] + df['Lesoes_Desconhecidas_Passageiros'] + df['Lesoes_Desconhecidas_Terceiros']
df['Total_Ilesos]'] = df['Ilesos_Tripulantes'] + df['Ilesos_Passageiros']

colunas_lesoes_totais = [
    'Total_Lesoes_Fatais', 'Total_Lesoes_Graves', 'Total_Lesoes_Leves',
    'Total_Lesoes_Desconhecidas', 'Total_Ilesos'
]

Visualizando as alterações

In [ ]:
display(df.head())

In [ ]:
colunas_manter = ['Territorialidades', 'IDHM 2012', 'IDHM 2013', 'IDHM 2014', 'IDHM 2015', 
                  'IDHM 2016', 'IDHM 2017', 'IDHM 2018', 'IDHM 2019', 'IDHM 2020', 
                  'IDHM 2021', 'IDHM 2022', 'IDHM 2023', 'IDHM 2024']

# ==========================================================
# PARTE 1: Transforma as colunas de anos e faixas em linhas
# ==========================================================

# A) Derretendo os valores numéricos do IDH
colunas_idh = ['Territorialidades', 'IDHM 2012', 'IDHM 2013', 'IDHM 2014', 'IDHM 2015', 
               'IDHM 2016', 'IDHM 2017', 'IDHM 2018', 'IDHM 2019', 'IDHM 2020', 
               'IDHM 2021', 'IDHM 2022', 'IDHM 2023', 'IDHM 2024']

df_idh_valores = pd.melt(
    df_idhm[colunas_idh], 
    id_vars=['Territorialidades'], 
    var_name='Ano', 
    value_name='IDH'
)
df_idh_valores['Ano'] = df_idh_valores['Ano'].str.replace('IDHM ', '').astype(int)


# B) Derretendo as faixas textuais de classificação
# CORREÇÃO: Adicionada a coluna 'Faixa_de_IDHM.12' para fechar os 13 anos (2012 a 2024)
colunas_faixas = ['Territorialidades', 'Faixa_de_IDHM', 'Faixa_de_IDHM.1', 'Faixa_de_IDHM.2', 'Faixa_de_IDHM.3', 
                  'Faixa_de_IDHM.4', 'Faixa_de_IDHM.5', 'Faixa_de_IDHM.6', 'Faixa_de_IDHM.7', 'Faixa_de_IDHM.8', 
                  'Faixa_de_IDHM.9', 'Faixa_de_IDHM.10', 'Faixa_de_IDHM.11', 'Faixa_de_IDHM.12']

df_faixas_texto = pd.melt(
    df_idhm[colunas_faixas], 
    id_vars=['Territorialidades'], 
    var_name='Coluna_Original', 
    value_name='Faixa_IDH'
)

# Mapeia a sequência exata de colunas de faixa para bater com os anos correspondentes (2012 a 2024)
mapeamento_anos = dict(zip(colunas_faixas[1:], range(2012, 2025)))
df_faixas_texto['Ano'] = df_faixas_texto['Coluna_Original'].map(mapeamento_anos)


# C) Juntando tudo lado a lado
df_idh_longo = pd.merge(
    df_idh_valores, 
    df_faixas_texto[['Territorialidades', 'Ano', 'Faixa_IDH']], 
    on=['Territorialidades', 'Ano']
)

# ==========================================================
# PARTE 2 E 3: Limpeza final e ajuste de nomes para o merge
# ==========================================================
df_idh_final = df_idh_longo[df_idh_longo['Territorialidades'] != 'BR'].copy()
df_idh_final.rename(columns={'Territorialidades': 'UF'}, inplace=True)

# Organiza a exibição visual final
df_idh_final = df_idh_final[['UF', 'Ano', 'IDH', 'Faixa_IDH']]

display(df_idh_final.head(10))
